### PR TITLE
Bug 2094196: Added webhook support for PowerVS platform

### PR DESCRIPTION
**Pull request by Karthik-K-N**
This PR will add validation and defaulting webhook support for PowerVS Platform

Sample Machine Spec can be found [here](https://github.com/openshift/machine-api-provider-powervs/blob/81e0c5c8fc2dc5b2f5ff2e32a37ff673ac9ba407/examples/machine-with-user-data.yaml)

### CODE REVIEW

File: pkg/webhooks/machine_webhook.go
Comment (mkumatag) [inline pkg/webhooks/machine_webhook.go:120]: @clnperez need your help deciding what should be the right values here for any worker ^^

File: go.mod
Comment (mkumatag) [inline go.mod:35]: why a separate entry here? can we include in the above block itself?

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: pkg/webhooks/machine_webhook.go
Comment (Karthik-K-N) [reply pkg/webhooks/machine_webhook.go:120]: @clnperez shall I go ahead with this https://github.com/openshift-powervs/installer/blob/2e17e5afefda5976366d71caa7cdee03fa42ceec/pkg/asset/machines/worker.go#L165  as default values.

Comment (clnperez) [reply pkg/webhooks/machine_webhook.go:120]: i would recommend the values the UPI folks have been using: https://github.com/ocp-power-automation/ocp4-upi-powervs/blob/7279803eed6566faf820bcac1334dc8e7f5ba761/compute-vars/medium.tfvars#L6

File: pkg/webhooks/machine_webhook.go
Comment (mkumatag) [inline pkg/webhooks/machine_webhook.go:119]:
```suggestion
	defaultPowerVSProcessor         = "0.5"
```

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: go.mod
Comment (Karthik-K-N) [reply go.mod]: Updated it in the latest commit

File: pkg/webhooks/machineset_webhook_test.go
Comment (mkumatag) [inline pkg/webhooks/machineset_webhook_test.go:234]: why are we modifying the existing testcase?

File: pkg/webhooks/machineset_webhook_test.go
Comment (mkumatag) [inline pkg/webhooks/machineset_webhook_test.go:341]: why we are removing this block and also see lot of code commented,

File: pkg/webhooks/machineset_webhook_test.go
Comment (Karthik-K-N) [reply pkg/webhooks/machineset_webhook_test.go]: Fixed it in the latest commit

File: pkg/webhooks/machineset_webhook_test.go
Comment (Karthik-K-N) [reply pkg/webhooks/machineset_webhook_test.go]: Fixed it in the latest commit

File: pkg/webhooks/machine_webhook.go
Comment (Karthik-K-N) [inline pkg/webhooks/machine_webhook.go:1416]: Machine network validation is done here

File: go.mod
Comment (JoelSpeed) [inline go.mod:15]: This dependency then becomes a dependency of all other providers, which makes bumping the deps a nightmare.

If you want defaulting for the powervs provider spec, it must first be moved into `openshift/api` with the other platforms

**Review by JoelSpeed — COMMENTED** [review]
Codewise I don't have much to comment on, but I don't think we can merge this as is because of the new dependency it introduces.

PowerVS is a new platform for 4.11 right? Has your API had any sort of API review? It would be good to get this done, and, ideally, migrate it into openshift/api

File: go.mod
Comment (Karthik-K-N) [reply go.mod:15]: Got it, I hope you mean adding the [`PowerVSMachineProviderConfig`](https://github.com/Karthik-K-N/machine-api-provider-powervs/blob/f8d06b4337415daa86e56295c297fef0e615575d/pkg/apis/powervsprovider/v1alpha1/powervsmachineproviderconfig_types.go#L31) to [machine/v1beta1](https://github.com/openshift/api/tree/master/machine/v1beta1)

Comment (mkumatag) [reply go.mod:15]: @Karthik-K-N lets submit a code for the review :)

Comment (JoelSpeed) [reply go.mod:15]: Please make sure to review the API conventions I linked in my previous comment and try to ensure your code meets the requirements before you submit, it will make the process easier

**Comment by JoelSpeed** [comment]
I had a conversation with David about the UX here and there are a few validations I would like to see

- Validate the Machine Type is one known to us, if it is not, send a warning to the user that we do not recognise the machine type
- Validate the CPU and Memory restrictions based upon the machine type. The limits are documented by IBM, eg S922 Machines can have up to 22 cores and 4TB of memory, we can validate that and return a good error message to the user at config time rather than at runtime

**Comment by Karthik-K-N** [comment]
I have updated this PR to use Power VS apis from opeshift/api repo and This PR is ready for a review

File: pkg/webhooks/machine_webhook.go
Comment (JoelSpeed) [inline pkg/webhooks/machine_webhook.go:161]: Nit, I would Capitalise the E to make it clear it's something new
```suggestion
	powerVSSystemTypeE880           = "e880"
	powerVSSystemTypeE980           = "e980"
```

File: pkg/webhooks/machine_webhook.go
Comment (JoelSpeed) [inline pkg/webhooks/machine_webhook.go:1507]: This depends on the proc type right? We can't set a 0.5 value for some of the proc type values?

File: pkg/webhooks/machine_webhook.go
Comment (JoelSpeed) [inline pkg/webhooks/machine_webhook.go:1508]: We don't allow a zero value so we can assume that if it's set to zero, we are going to default it, I think this is fine as is

File: pkg/webhooks/machine_webhook.go
Comment (JoelSpeed) [inline pkg/webhooks/machine_webhook.go:1605]: What about regex type? Why isn't that listed here?

**Review by JoelSpeed — COMMENTED** [review]
Not sure how easy this is to do now, but it would be much easier to review if the machinev1 to machinev1beta1 import rename was a separate commit at the beginning of the chain, is that possible?

**Comment by Karthik-K-N** [comment]
> Not sure how easy this is to do now, but it would be much easier to review if the machinev1 to machinev1beta1 import rename was a separate commit at the beginning of the chain, is that possible?

I think that can be done, I will give a try, But how about a separate PR with api bump and machinev1beta1 changes.

**Comment by Karthik-K-N** [comment]
> > But how about a separate PR with api bump and machinev1beta1 changes.
> 
> That would be even better!

I was able to update it in the same PR, Also i have addressed all your comments.

**Comment by Karthik-K-N** [comment]
/hold
hold until respective PR for api update is merged to Power VS mapi:https://github.com/openshift/machine-api-provider-powervs/pull/17

File: pkg/webhooks/machine_webhook.go
Comment (JoelSpeed) [reply pkg/webhooks/machine_webhook.go:163]: This comment still needs resolving

**Review by JoelSpeed — COMMENTED** [review]
One nit, but otherwise LGTM

File: pkg/webhooks/machine_webhook.go
Comment (Karthik-K-N) [reply pkg/webhooks/machine_webhook.go]: I missed it earlier, Updated it now

**Comment by Karthik-K-N** [comment]
> /approve cancel
> 
> The vendor directory here looks inconsistent, why are there vendor changes without equivalent go.mod and go.sum entries? I can see that we are pulling in API types from the machine-api-provider-powervs repository, that is not allowed. You need to move your API types through to the openshift/api repo before these webhooks can be introduced
> 
> This will have to be done in 4.12

Seems like inconsistencies due to rebase effort, I have udpated it now.

**Review by JoelSpeed — COMMENTED** [review]
Please drop the two vendor commits as they cancel each other out, otherwise LGTM. Will ask for another pair of eyes over it before we merge

**Comment by Karthik-K-N** [comment]
> Please drop the two vendor commits as they cancel each other out, otherwise LGTM. Will ask for another pair of eyes over it before we merge

Dropped the vendor commits

**Review by Fedosin — APPROVED** [review]
/lgtm

**Comment by JoelSpeed** [comment]
/override ci/prow/e2e-vsphere-serial 

This change doesn't affect vsphere

**Pull request merged**
